### ISSUE TITLE
paddleocr可以直接保存训练中数据增强的图片吗？是不能使用config文件而必须用底层paddle来做吗？

### ISSUE THREAD

**Issue by OpenOneV** [PaddlePaddle/PaddleOCR#10130]
提问
2023-06-09 18:17:04
paddleocr如何保存数据增强之后的图片，在训练的过程中？

回答
2023-06-09 18:17:04
PaddleOCR可以使用train.py中的transform_kwargs参数保存数据增强后的图片。下面是一个实例代码：

import paddleocr

transform_kwargs = {'image_distort_strategy': 'hue',
                    'hue_prob': 0.5,
                    'hue_range': (-30, 30)}

ocr = paddleocr.OCR(rec_algorithm='crnn', transform_kwargs=transform_kwargs)

# 在训练过程中使用数据增强，并将增强后的图片保存
ocr.train(train_datasets=train_reader,
          train_params=train_params,
          eval_datasets=eval_reader,
          learning_rate=learning_rate,
          save_dir=model_save_dir,
          use_visualdl=True,
          visualdl_logdir=visualdl_log_dir)
在上述代码中，将transform_kwargs参数设置为{'image_distort_strategy': 'hue','hue_prob': 0.5,'hue_range': (-30, 30)}，表示使用色调扭曲（hue distortion）策略，并在图片色调范围内随机调整色调。当训练PaddleOCR时，通过在train()函数中传递train_reader和transform_kwargs参数，可以对训练集数据进行数据增强，并将增强后的图片保存下来。

提问
2023-06-09 18:21:15
如何保存数据增强之后的图片

回答
2023-06-09 18:21:16
PaddleOCR可以在定义DataLoader对象时使用transforms.Compose进行数据增强，并在训练过程中使用paddle.fluid.layers.py_reader()读取增强后的图片。下面是一个示例代码：

import os
import numpy as np
import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph import Linear
from paddle.fluid.dygraph import Conv2D
from paddle.fluid.dygraph import Pool2D
from paddle.fluid.dygraph import BatchNorm

transform = transforms.Compose([
    transforms.Resize((32, 100)),
    transforms.RandomHorizontalFlip(),  # 水平翻转
    transforms.ColorJitter(brightness=0.4, contrast=0.4, saturation=0.4),  # 修改亮度、对比度、饱和度
    transforms.ToTensor(),  # 转化为张量
])

with fluid.dygraph.guard():
    # 训练样本目录
    train_folders = ['data/0', 'data/1', 'data/2', 'data/3', 'data/4', 'data/5',
                     'data/6', 'data/7/', 'data/8', 'data/9']
    train_files = []
    for folder in train_folders:
        files = os.listdir(folder)
        for file_name in files:
            file_path = os.path.join(folder, file_name)
            train_files.append(file_path)

    np.random.shuffle(train_files)

    # 定义数据分布式读取器（使用多线程异步方式加速数据读取）
    train_loader = fluid.io.DataLoader.from_generator(capacity=5)
    train_loader.set_sample_generator(
        paddle.batch(paddle.reader.shuffle(paddle.dataset.\
            mnist.train(), buf_size=500),
            batch_size=128), places=fluid.CPUPlace())
    
    # 定义神经网络结构
    class OCR(fluid.dygraph.Layer):
        def __init__(self):
            super(OCR, self).__init__()
            self.conv1 = Conv2D(num_channels=1, num_filters=32, filter_size=3, stride=1, padding=1, act='relu')
            self.pool1 = Pool2D(pool_size=2, pool_stride=2, pool_type='max')
            self.bn1 = BatchNorm(num_channels=32, act='relu')
            
            self.conv2 = Conv2D(num_channels=32, num_filters=64, filter_size=3, stride=1, padding=1, act='relu')
            self.pool2 = Pool2D(pool_size=2, pool_stride=2, pool_type='max')
            self.bn2 = BatchNorm(num_channels=64, act='relu')
            
            self.fc1 = Linear(input_dim=64*8*25, output_dim=256, act='relu')
            self.fc2 = Linear(input_dim=256, output_dim=10)

        def forward(self, inputs):
            x = inputs
            x = self.conv1(x)
            x = self.pool1(x)
            x = self.bn1(x)

            x = self.conv2(x)
            x = self.pool2(x)
            x = self.bn2(x)

            x = fluid.layers.reshape(x, shape=[x.shape[0], -1])
            x = self.fc1(x)
            x = self.fc2(x)

            return x

    # 定义优化器和学习率衰减策略
    with fluid.dygraph.guard():
        model = OCR()
        opt = fluid.optimizer.AdamOptimizer(learning_rate=1e-3, parameter_list=model.parameters())

        for epoch in range(100):
            for i, data in enumerate(train_loader()):
                image, label = data
                image = transform(image)  # 进行数据增强

                # 转化为浮点型，并调整形状
                image = fluid.layers.cast(image, dtype='float32')
                image = fluid.layers.reshape(image, shape=[-1, 1, 32, 100])
                label = fluid.layers.cast(label, dtype='int64')

                # 前向传播计算损失值
                predict = model(image)
                loss = fluid.layers.cross_entropy(predict, label)
                avg_loss = fluid.layers.mean(loss)

                # 反向传播更新参数
                avg_loss.backward()
                opt.minimize(avg_loss)
                model.clear_gradients()

            # 打印日志
            if epoch % 10 == 0:
                print('epoch: %05d, loss=%.5f' % (epoch, loss.numpy()))   
在上述代码中，将使用transforms.Compose函数对训练数据进行数据增强。例如，增加随机水平翻转策略、修改亮度、对比度、饱和度等。train_loader为一个数据分布式读取器，使用多线程异步方式加速数据读取，每次读取一个batch的训练样本。使用多线程异步方式加速数据读取，每次读取一个batch的训练样本。在每个batch的数据增强和处理完成后，将数据转化为张量，并调整张量大小，以便能够输入到神经网络中进行训练。在训练过程中，使用paddle.fluid.layers.py_reader()读取数据，可以自动将数据分配到各个GPU卡中，并且对训练数据进行数据增强。

**Comment by OpenOneV** [comment]
回答
2023-06-09 18:25:01
是的，PaddleOCR可以直接保存训练中的数据增强的图片。但是，这需要在代码中手动添加保存代码，而不是通过使用config文件来实现。可以使用PaddlePaddle框架中的Image.save()方法，在训练中将图片保存到本地。以下是一个保存图像的示例代码片段：

import paddle
from paddle.vision.transforms import transforms

# 定义预处理函数
transform = transforms.Compose([
    transforms.Resize(size=(224, 224)),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

# 获取一批数据并保存到本地
for data in dataloader:
    images, labels = data
    for i in range(images.shape[0]):
        img = transform(images[i])
        paddle.to_tensor(img).save("my_images/{}.jpg".format(i))
这个例子中，我们首先定义了一个预处理函数和一个dateloader来获取图像数据。在训练循环中，我们循环遍历每张图片，并使用预处理函数将其转换为张量，并使用paddle.to_tensor(img).save()方法将其保存在本地磁盘上。

需要注意的是，这种方法可能会导致磁盘空间占用较大，因为每个epoch会保存所有的数据增强后的图片。如果需要限制磁盘空间，则需要在代码中设置删除策略。
